neqo-transport/src/connection/mod.rs

```diff
@@ -1267,24 +1302,27 @@ impl Connection {
             debug_assert!(self.crypto.states.rx_hp(self.version, cspace).is_some());
             for saved in self.saved_datagrams.take_saved() {
                 qtrace!([self], "input saved @{:?}: {:?}", saved.t, saved.d);
-                self.input(&saved.d, saved.t, now);
+                self.input((&saved.d).into(), saved.t, now);
             }
         }
     }
 
     /// In case a datagram arrives that we can only partially process, save any
     /// part that we don't have keys for.
-    fn save_datagram(&mut self, cspace: CryptoSpace, d: &Datagram, remaining: usize, now: Instant) {
-        let d = if remaining < d.len() {
-            Datagram::new(
-                d.source(),
-                d.destination(),
-                d.tos(),
-                &d[d.len() - remaining..],
-            )
-        } else {
-            d.clone()
-        };
+    fn save_datagram(
+        &mut self,
+        cspace: CryptoSpace,
+        d: Datagram<&[u8]>,
+        remaining: usize,
+        now: Instant,
+    ) {
+        let d = Datagram::new(
+            d.source(),
+            d.destination(),
+            d.tos(),
+            d[d.len() - remaining..].to_vec(),
+            Some(d.segment_size()),
+        );
```

Comment:
Maybe you could implement a `to_owned()` for `Datagram` rather than do it this way.